src/Workspaces/Core/MSBuild/Rpc/RpcClient.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Rpc;
+
+/// <summary>
+/// Implements the client side of the RPC channel used to communicate with the build host, which is using <see cref="RpcServer"/>.
+/// </summary>
+/// <remarks>
+/// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+/// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+/// are out. If at some point there is a standard RPC mechanism exposed in .NET or Source Build, we should delete this and use that instead.
+/// </remarks>
+internal sealed class RpcClient
+{
+    private readonly TextWriter _sendingStream;
+    private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+    private readonly TextReader _receivingStream;
+
+    private readonly ConcurrentDictionary<int, (TaskCompletionSource<object?>, System.Type? expectedReturnType)> _outstandingRequests = new();
+    private volatile int _nextRequestId = 0;
+
+    private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+    public RpcClient(Stream sendingStream, Stream receivingStream)
+    {
+        _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+        _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
+    }
+
+    public event EventHandler? Disconnected;
+
+    public void Start()
+    {
+        // We'll start this and let it run until Shutdown has been called.
+        Task.Run(async () =>
+        {
+            string? line;
+            while ((line = await _receivingStream.TryReadLineOrReturnNullIfCancelledAsync(_shutdownTokenSource.Token).ConfigureAwait(false)) != null)
+            {
+                var response = JsonConvert.DeserializeObject<Response>(line);
+
+                Contract.ThrowIfNull(response);
+
+                Contract.ThrowIfFalse(_outstandingRequests.TryRemove(response.Id, out var completionSourceAndExpectedType), $"We got a response for request ID {response.Id} but that was already completed.");
```

Comment:
are these exceptions logged anywhere?

Reply:
I'm trying to think what the best opportunity would be. I can't just easily throw the exception from somewhere, and since this is an API not running in something like VS I can't just do our usual reporting that way. Thoughts?

Reply:
woudl like the body of this while to be in its own helper.

Reply:
Wondering if we want to somehow hookup NFW, or have a specific file log or something.
We can pass in like the vscode extension logs folder and write something there if needed.

Reply:
As mentioned: server exceptions go to the client if they were a part of a client call, otherwise go to stderr which we log. We otherwise don't have a place we can really log some of these, since this at such a low level we won't have NFW in practical scenarios. For example, if it's an end-user app we don't have NFWs at all, and if this is being used in VS Code it'd be Blue-only so we won't have telemetry either.